Add progress bars for categories and goals

tbuist opened the pull request:
Fixes #2965 

This change introduces visual bars to indicate the status of a particular category for that month. 

For standard expense categories, green, red, and gray indicate the value you have remaining/overspent in that category.

For categories with a #goal template, the bar uses blue to indicate the progress.

![image](https://github.com/user-attachments/assets/ffdd0e0c-cf65-4d26-bdd2-edeea2041b27)


<!-- Thank you for submitting a pull request! Make sure to follow the instructions to write release notes for your PR — it should only take a minute or two: https://github.com/actualbudget/docs#writing-good-release-notes -->

youngcw commented:
Do you have to do something to turn on the progress bars?

What will happen when viewing multiple months?

tbuist commented:
> Do you have to do something to turn on the progress bars?

Yes, I added a global preference in the Themes section. I saw the documentation recommends against adding configuration for small UI tweaks, so I'm happy to hear feedback on if this is big enough to warrant a config option, or if it belongs somewhere else like a feature flag (or on by default, but I didn't want to presume).
![image](https://github.com/user-attachments/assets/ad5b6feb-0910-4b66-87d7-3b6629ff6e9f)

 
> What will happen when viewing multiple months?

The bars only appear in the sidebar that hosts the category titles, and only consider budgetting/spending in first month that is being looked at. Any thoughts about that? Two examples:
![image](https://github.com/user-attachments/assets/65c1853d-aa9b-4f78-94b8-0aa0b6686855)
![image](https://github.com/user-attachments/assets/e4541c77-b3cb-4d26-81f7-65b9a86e4c61)

Teprifer commented:
> Do you have to do something to turn on the progress bars?
> 
> What will happen when viewing multiple months?

Tick box on the settings page by the theme selector.

Multi month looks like this, with the bars calculated for the left most month: 
![image](https://github.com/user-attachments/assets/8740bad5-fd31-4f6d-8b2f-47d929c5dec8)

Multi-month presents a tricky UI challenge, probably simplest might be to have it only calculated for the current month, and all hidden when the current month isn't on screen.

My initial thoughts, were the bars look cramped.

Categories without any goal, budgeted, or spent, still have the bar space reserved.

Not sure about the mixed use of a bar to represent goal progress and budgeted vs spent as it's budgeting with goals, and usually goals are paired with templates, but the bar doesn't incorporate those because that way lies madness. 

I can see a category bar will go red because an expense was recorded against it, indicator a 'negative' or bad thing, when really it's fine because the category balance is still positive, and the money for it had been budgeted in a prior month. So a little misleading?

Might be worth taking a look at #3424 which attempted to incorporate the multi-month aspects.

tbuist commented:
> Multi-month presents a tricky UI challenge, probably simplest might be to have it only calculated for the current month, and all hidden when the current month isn't on screen.

This seems reasonable and intuitive to me.
 
> My initial thoughts, were the bars look cramped.
>
> Categories without any goal, budgeted, or spent, still have the bar space reserved.

Currently the color would be empty at least, so the bar still shows but doesn't draw as much attention.
![image](https://github.com/user-attachments/assets/f4a4d805-bab3-47de-9512-8ba5cfd7de2f)

> Not sure about the mixed use of a bar to represent goal progress and budgeted vs spent as it's budgeting with goals, and usually goals are paired with templates, but the bar doesn't incorporate those because that way lies madness.

Fair point.

> I can see a category bar will go red because an expense was recorded against it, indicator a 'negative' or bad thing, when really it's fine because the category balance is still positive, and the money for it had been budgeted in a prior month. So a little misleading?

Also a fair point. 

> Might be worth taking a look at #3424 which attempted to incorporate the multi-month aspects.

Wow, I'm not sure how I missed that entire thread when looking for other attempts at this problem. After skimming, this does seem like a more complex attempt than what I intended mine to be. Thinking more deeply about what I envisioned these bars representing, I think I was targeting only the two most basic use cases:

1. For expense/template categories in one particular month, how much did I intend to spend, and how much did I actually spend?
2. For goal categories in one particular month, how much did I intend to save, and how much do I now have saved in total as a result?

I think those cover the intended use cases and avoid entirely the complexity of communicating single- and multi-month statuses for a category at the same time, but obviously does so by limiting the amount of information being shown. That might be a philosophical difference, but I will mull on what multi-month support would need to look like.

lelemm commented:
I like the idea, but I find it too distracting, not sure if the colors are too intense, or the colors doesn't fit to the overall theme. also, I agree with @youngcw.

> I think the progress bars really should be per month and not next to the categories, if we are going to add them. Also this currently doesn't have any indication of underfunded categories which will need added too.

tbuist commented:
Here's a version that moves the bars out of the category column and also deemphasizes them a little bit by just shrinking the height - thoughts on this?
![image](https://github.com/user-attachments/assets/74ff376d-8179-436d-ae21-a2d42f026a1a)

Here's the same height in their original location
![image](https://github.com/user-attachments/assets/53fb3280-f5ef-4c68-9300-d0723c379860)

My thought is that the shorter the horizontal length of the bar, the quicker I'm able to glean what the state of the category is without parsing the actual amounts.

lelemm commented:
The smaller height makes it way better!
Try to use the color variables from the themes:
![image](https://github.com/user-attachments/assets/f81ec00c-189b-4098-a460-274487bf7e58)

what if the current month shows the bar, but for other months you have to hover the month container?

tbuist commented:
How do we feel about this? I'm just hooking into the same hover styling behavior that already exists on ExpenseCategoryMonth.
![output](https://github.com/user-attachments/assets/4c6ec90c-da35-439a-85a6-6de46ba4dbf5)

Alternatively, I can explore highlighting every category in the month when any part of the month component is hovered. I didn't see a clean way to do this with the way the components are structured, but I can mess around some more if we want to see that.

Separately, I've switched to using the report colors.
```
const ColorDefUnderBudgetRemaining = theme.reportsGreen; 
const ColorDefUnderBudgetSpent = theme.reportsGray; 
const ColorDefOverBudgetSpent = theme.reportsGray; 
const ColorDefOverBudgetOverSpent = theme.reportsRed;
const ColorDefGoalRemaining = theme.reportsLabel; 
const ColorDefGoalSaved = theme.reportsBlue; 
const ColorDefEmpty = ''; // No color for default
```

tbuist commented:
One more to consider:
![output](https://github.com/user-attachments/assets/cb4c5e49-47c9-425a-ac7b-6d1071f2e147)

diegofede commented:
This would be really nice to have! Thanks for the proposal, but I would prefer it how another tool (ynab) does it:

There is a container, so the bar doesn't really go edge to edge, this would also simplify the view for multiple months so the bars do not touch. The coloring is very simple too, only 3 colors: green, striked green, striked red. Where striked means spent.

This accounts for all the possibilities:

- Available to spend: green
- Spent within budget: striked green
- Overspent: striked red
- Goals: same rules as above but the left to reach the goal is grey (which is the color of the container).

This will simplify with less colors and people coming from the other tool will feel more at home.

youngcw commented:
I'm not sure on the hover thing.  If you do add it, I think the hover to show all categories option (your second option) makes the most sense.  For starters you could just leave the progress bars on for all months that have goal data.  Then we could see if its distracting to have non-current month progress bars.

I do appreciate the thin lines.  Its not too obtrusive.  I always thought the YNAB progress bars were way too big and made the budget table super cluttered.  Back when I used YNAB I disabled the progress bars the second they got added to the app because they were annoying.  The way you have the PR, I might actually try it. 🙂 

It probably would be best to move the enable option into the "Category" menu.  Same place as the "Toggle Hidden Categories" option.

tbuist commented:
> I'm not sure on the hover thing. If you do add it, I think the hover to show all categories option (your second option) makes the most sense. For starters you could just leave the progress bars on for all months that have goal data. Then we could see if its distracting to have non-current month progress bars.

Something like this? This made me realize that the non-goal progress bars are not of much use for non-current month visuals. They are most useful for communicating how much you have left to spend (although I suppose we still need to confirm whether this should be relative to the budgeted amount or the category balance - currently it's the budgeted amount), but 1) you can't change spending decisions for past months, and 2) spending for future months is always 0, even with scheduled transactions, so these would always be green under the current behavior.
![output](https://github.com/user-attachments/assets/59fc2310-0826-46ed-b2c2-07d397666d1d)

> Im not sold on having progress bars on categories without goals. I feel like that detracts from the usefulness of the categories with templates if all categories have a progress bar looking for attention.

Fewer colors helps.
![output](https://github.com/user-attachments/assets/ed62f23d-1b2e-4fbe-9a99-84b8e60f44cd)


> It probably would be best to move the enable option into the "Category" menu. Same place as the "Toggle Hidden Categories" option.

Good idea, I'll move it here.

lelemm commented:
I like how the other months are faded.
The thin bar looks better.

how does it look like if the progress bar is just under the balance column?

tbuist commented:
Here's just under the balance column. If we went this route, do you think the bar ratios would need to reflect spending relative to the balance, as opposed to spending relative to budgeted?

I do think this is easier to read though.

![image](https://github.com/user-attachments/assets/f951c481-c4f4-40e8-ac1d-89adf614a39f)

lelemm commented:
> Here's just under the balance column. If we went this route, do you think the bar ratios would need to reflect spending relative to the balance, as opposed to spending relative to budgeted?
> 
> I do think this is easier to read though.
> 
I like this version.

If its under the balance, I would prefer the progress bar to be the opposite. Right now you are showing from 100% to 0% (showing how much budgeted you still have based on spent). 
If you put this current logic under budgeted, it gets more intuitive.

Could you update your branch with the current code so we can play with it too?

tbuist commented:
Yep, just pushed everything. I moved the bar under the 'budgeted' column since that does make more sense with the current logic, like you mentioned. Let me know what you think.

lelemm commented:
@tbuist , I played a little with your code, I liked it this way:

![image](https://github.com/user-attachments/assets/4e78f673-9af3-4697-b541-55e1960bdc15)
![image](https://github.com/user-attachments/assets/c7642188-0c86-4a12-a0d3-28c47bf8d9e9)

what do you think?

Check my changes here: https://github.com/tbuist/actual/compare/tbuist-category-ui-bars...lelemm:actual:bars?expand=1
I had to merge it with the current master because your branch was not working for me (some LRUCache errors)

diegofederico1 commented:
> @tbuist , I played a little with your code, I liked it this way:
> 
> ![image](https://private-user-images.githubusercontent.com/15043428/417594521-4e78f673-9af3-4697-b541-55e1960bdc15.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2NzU1NzcsIm5iZiI6MTc0MDY3NTI3NywicGF0aCI6Ii8xNTA0MzQyOC80MTc1OTQ1MjEtNGU3OGY2NzMtOWFmMy00Njk3LWI1NDEtNTVlMTk2MGJkYzE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI3VDE2NTQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwMjk0OWFhMGViYmMxODhlMjUyNTlmNzY3NWM4OWEwNjVkYjIxNGI3NWUzZGVjYjkzYTNiYTkyZDE4NTY4NTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.4rcYgyxd5o8TqyDcl_3fV0IWpysWtlJzFV4AD0Fl4Bg) ![image](https://private-user-images.githubusercontent.com/15043428/417596046-c7642188-0c86-4a12-a0d3-28c47bf8d9e9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2NzU1NzcsIm5iZiI6MTc0MDY3NTI3NywicGF0aCI6Ii8xNTA0MzQyOC80MTc1OTYwNDYtYzc2NDIxODgtMGM4Ni00YTEyLWEwZDMtMjhjNDdiZjhkOWU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI3VDE2NTQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNmYTVmMWNhYTMxNWQzNTY1ZTNjMDI4Y2JjYjU3ZGU3MjBiZjY5NjAxZDQ5YTFkZmRhYjllMzFiOWVkOTIxNjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.mP6qTfi1T83Z7-OP1LLa7KZ7mshI55chgK7qRs8DBHs) ![image](https://private-user-images.githubusercontent.com/15043428/417688695-9e168400-4c1c-43b8-a92d-dd409c0e0cd1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2NzU1NzcsIm5iZiI6MTc0MDY3NTI3NywicGF0aCI6Ii8xNTA0MzQyOC80MTc2ODg2OTUtOWUxNjg0MDAtNGMxYy00M2I4LWE5MmQtZGQ0MDljMGUwY2QxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI3VDE2NTQzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYxYTZjNDk4NGNmYzZkZGNjNWQ3ZDk2ZDcwMTIxYmM4YjMzMjlhNDc1ZjAwNWU0YjgzOTBkY2FiYmYyN2RkNzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.uEP9HdJ9IxP0LNJhs2FsE9cqR7UMeWza3G78Pq8dR78)
> 
> what do you think?
> 
> Check my changes here: https://github.com/tbuist/actual/compare/tbuist-category-ui-bars...lelemm:actual:bars?expand=1 I had to merge it with the current master because your branch was not working for me (some LRUCache errors)

I really like how it looks! But how would a category not yet budgeted look like? If it is the bar background color it would look the same as budgeted=spent, which is not too bad but could make it harder to spot non budgeted categories, which should be one of the advantages of having visual cues.

We can analyze the different scenarios:

Good

- Budgeted and no spending
- Budgeted and partially spent
- Budgeted and fully spent

Bad

- Budgeted and overspent
- No budget and no spending 
- No budget and spending

I think all of this scenarios should be visually obvious via a quick glance. This should be possible to achieve by having the spent within budget as a shade of green, the no spent part would be green. The overspent part would be red. Background color would be gray, like you already you used, which would represent no activity, no budget no spending, nothing. A combination of these would cover all scenarios.

Opinions on the color palette may vary, but I think we need to start with an overview of all scenarios first. Speaking of which, Goals:

Goals are more complicated because they are not really final on the current implementation but it would look something like this

Good
Budgeted equal or more than enough to reach the target for the month

Bad
Not budgeted enough to reach target for the month

Goals would need more thinking, but currently on a train and my stop is soon.

deathblade666 commented:
Firstly, this is looking great excellent job!!

But just a note and it maybe something that could be clarified in the docs. To me without initially making the connection that the progress bars are designed to visualize what's been budgeted and not the category balance throw me off. After looking at it more it made sense what it was showing but at first glance was a bit confusing. 

If it was requested to be based on budgeted amount rather than balance then I admit I have not read the whole thread. Just wanted to provide a potentially more outside perspective.

youngcw commented:
> @tbuist , I played a little with your code, I liked it this way:
> 
> ![image](https://private-user-images.githubusercontent.com/15043428/417594521-4e78f673-9af3-4697-b541-55e1960bdc15.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2ODI3NjQsIm5iZiI6MTc0MDY4MjQ2NCwicGF0aCI6Ii8xNTA0MzQyOC80MTc1OTQ1MjEtNGU3OGY2NzMtOWFmMy00Njk3LWI1NDEtNTVlMTk2MGJkYzE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI3VDE4NTQyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBiODA2NjM1ZDlmZGU3ZDI0Zjc4OTUwMmI2ZGE3N2Y5YmRjNTU1YjQ4YjExYWFmZWRmZjEyYmRjNjgyNzE3YmYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NsWYBlg9bv12-3vYUSp6PFn-5QXEhI9QNPIGOqKmzP0) ![image](https://private-user-images.githubusercontent.com/15043428/417596046-c7642188-0c86-4a12-a0d3-28c47bf8d9e9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2ODI3NjQsIm5iZiI6MTc0MDY4MjQ2NCwicGF0aCI6Ii8xNTA0MzQyOC80MTc1OTYwNDYtYzc2NDIxODgtMGM4Ni00YTEyLWEwZDMtMjhjNDdiZjhkOWU5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI3VDE4NTQyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk3MzVjZmYwYTAwNTcwNmZhNjA2NmE0ZjEyMDc4MjFhMDIxNGEyNWNjMGVlNTNmODhiMTNhMTQ2MjM3NTcyNGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Ug9j9TcGSBgFFGnsXz7pUo3dZtkQ8HClz-Q8TxiXokk) ![image](https://private-user-images.githubusercontent.com/15043428/417688695-9e168400-4c1c-43b8-a92d-dd409c0e0cd1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2ODI3NjQsIm5iZiI6MTc0MDY4MjQ2NCwicGF0aCI6Ii8xNTA0MzQyOC80MTc2ODg2OTUtOWUxNjg0MDAtNGMxYy00M2I4LWE5MmQtZGQ0MDljMGUwY2QxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI3VDE4NTQyNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdmMGY2NWQ0YzY4Zjg5YjI2NjU0M2UyYzkwNGRkNzg5MmJiYjE0NThmYWE1Y2U2ZDI3NmY3ZDg3ZjM2NjUyMzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.vS2_USpdXDkgVp9WbV-awMkwBvq3JLg0w9ODIazQX-s)
> 
> what do you think?
> 
> Check my changes here: https://github.com/tbuist/actual/compare/tbuist-category-ui-bars...lelemm:actual:bars?expand=1 I had to merge it with the current master because your branch was not working for me (some LRUCache errors)

would this have the same confusion with `#goal` that is based off of balance instead of budgeted?

youngcw commented:
No matter which side you choose there will be templates that are based on the other column.  So it seems like if you only use one column of space for the progress it will cause confusion.  No matter which side is picked.

lelemm commented:
> No matter which side you choose there will be templates that are based on the other column. So it seems like if you only use one column of space for the progress it will cause confusion. No matter which side is picked.

what if we split the progress bar into two? do you think its still confusing? the current implementation of  @tbuist it tries to show goals and budgeted all at the same progress bar.
we could have a simple budgeted vs spent under budgeted value and a goal progress bar under balance

![image](https://github.com/user-attachments/assets/d1df98da-52ae-4ae4-9fdb-33434268e796)
(fake code here)

deathblade666 commented:
> > No matter which side you choose there will be templates that are based on the other column. So it seems like if you only use one column of space for the progress it will cause confusion. No matter which side is picked.
> 
> what if we split the progress bar into two? do you think its still confusing? the current implementation of @tbuist it tries to show goals and budgeted all at the same progress bar. we could have a simple budgeted vs spent under budgeted value and a goal progress bar under balance
> 
> ![image](https://private-user-images.githubusercontent.com/15043428/417792477-d1df98da-52ae-4ae4-9fdb-33434268e796.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDA2OTU2MjQsIm5iZiI6MTc0MDY5NTMyNCwicGF0aCI6Ii8xNTA0MzQyOC80MTc3OTI0NzctZDFkZjk4ZGEtNTJhZS00YWU0LTlmZGItMzM0MzQyNjhlNzk2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjI3VDIyMjg0NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRmNGY3OGM1OWY1YmRiNGNmNTczZDEyYjNiZTVlNzJmM2U0ZjQwMjA4MWY4MjdjNDEzZDg0NDhiNzVkNTc3YjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YyiGMX1M5I2-qKaSOkgtiDYiidSXrrC_np48tWtKGak) (fake code here)

FWIW, to me, this approach seems less confusing overall. Just a thought on adding progress bars for goals, would it make more sense to move them to the existing goal tooltip?

tbuist commented:
> > No matter which side you choose there will be templates that are based on the other column. So it seems like if you only use one column of space for the progress it will cause confusion. No matter which side is picked.
> 
> what if we split the progress bar into two? do you think its still confusing? the current implementation of @tbuist it tries to show goals and budgeted all at the same progress bar. we could have a simple budgeted vs spent under budgeted value and a goal progress bar under balance

My thought would be that in both cases (budgeted vs. spent, as well as goal progress), the _budgeted_ amount is a component of the thing we're trying to communicate. That is, in both cases, I am interested in the budgeted amount  for how much is remaining or for the impact it's having on the goal. I'm not sure I like multiple bars in one line.

> @tbuist , I played a little with your code, I liked it this way:

I do like the narrowed width bars that you have here.

diegofederico1 commented:
Super confusing, the bars should simplify not complicate reading the table, one bar per category should be enough.

tbuist commented:
I settled on the following after messing around with a few of the suggestions in this thread. I'll submit it for review.
![image](https://github.com/user-attachments/assets/b77d0605-3020-491c-bc36-8fe05863d04f)

youngcw commented:
The bars feel short.  I think they should either be the full length of the cell input box, or the full table width (budgeted through to balance)

tbuist commented:
> The bars feel short. I think they should either be the full length of the cell input box, or the full table width (budgeted through to balance)

Updated to be the full length of the cell input box.

abhiy98 commented:
Will this work on mobile?

tbuist commented:
It does not, at least currently. I'll defer to maintainers if they want it added.
![image](https://github.com/user-attachments/assets/964aead4-ff75-40e0-b0a0-774862c9529b)

youngcw commented:
I think adding this to mobile would be good.  


Its confusing that the progress bars don't match the template status.  I think that if there is a template, the template should be what the goal is based off of, not the spending.  Right now the spending is taking priority over the template status.  This causes the situation where the template is happy, but the progress bar is red.  I have this on my budget with a yearly bill.  I spent way more then was budgeted this month, but it was all funds that were already in the category.  The template status is green, but the progress bar is red and reports that I'm overspent.

JonoSK commented:
Could there be one bar under Budgeted to indicate progress on a #template (ie have you contributed enough this month), then a separate bar under Balance to indicate progress on a #goal (and only show if there is a #goal for that category)?

tbuist commented:
> Its confusing that the progress bars don't match the template status. I think that if there is a template, the template should be what the goal is based off of, not the spending. Right now the spending is taking priority over the template status. This causes the situation where the template is happy, but the progress bar is red. I have this on my budget with a yearly bill. I spent way more then was budgeted this month, but it was all funds that were already in the category. The template status is green, but the progress bar is red and reports that I'm overspent.

Would this only apply to templates with a ['by' type](https://actualbudget.org/docs/experimental/goal-templates/#by-type)? I'd agree, it should probably be treated more like a #goal in that case.

youngcw commented:
> > Its confusing that the progress bars don't match the template status. I think that if there is a template, the template should be what the goal is based off of, not the spending. Right now the spending is taking priority over the template status. This causes the situation where the template is happy, but the progress bar is red. I have this on my budget with a yearly bill. I spent way more then was budgeted this month, but it was all funds that were already in the category. The template status is green, but the progress bar is red and reports that I'm overspent.
> 
> Would this only apply to templates with a ['by' type](https://actualbudget.org/docs/experimental/goal-templates/#by-type)? I'd agree, it should probably be treated more like a #goal in that case.

It would apply to any category that a user builds up funds, not just specific templates.

tbuist commented:
> > > Its confusing that the progress bars don't match the template status. I think that if there is a template, the template should be what the goal is based off of, not the spending. Right now the spending is taking priority over the template status. This causes the situation where the template is happy, but the progress bar is red. I have this on my budget with a yearly bill. I spent way more then was budgeted this month, but it was all funds that were already in the category. The template status is green, but the progress bar is red and reports that I'm overspent.
> > 
> > 
> > Would this only apply to templates with a ['by' type](https://actualbudget.org/docs/experimental/goal-templates/#by-type)? I'd agree, it should probably be treated more like a #goal in that case.
> 
> It would apply to any category that a user builds up funds, not just specific templates.

Ah, I misunderstood your first comment. Basically, you're suggesting the bar should represent the spending relative to the balance, which is the way we tell how "good" our budgeted amount is relative to any #templates we have.

Your example might be something like "#template 1200 by 2025-03 repeat every year". If I budget 100 but pay off 1200 in March, the bar should actually be all blue or all black (either an empty or a fulfilled #goal) - as opposed to saying I'm -1100 overspent, as the logic currently does. In April, the bar would reset in preparation for next year's payment. This makes sense.

That caveat for the bar type (being a goal ratio instead of a spent/budget ratio) only applies to the 'by' template type though, right? I think everything else is treated more as a month-to-month. 

The 'up to' template type would be another case where we care about the balance too. If the user rolled over extra funds, they probably expect their spending to be more than budgeted but equal to balance. Maybe that's a case where we replace 'budgeted' with 'balance' in our percentage spent calculations? This would still essentially be a spent/budget ratio though, where actual budget=budget+balance.

youngcw commented:
We have talked a bit behind the scenes.  We think having the progress bars be underneath the balance would be the best place.

youngcw commented:
> Basically, you're suggesting the bar should represent the spending relative to the balance, which is the way we tell how "good" our budgeted amount is relative to any #templates we have.

For starters we should start with the progress bars matching what the tooltip shows.  So that would not consider spending at all.  It would only look at what is the budgeted vs goal.  Then categories with long_goal set would look at balance vs goal.

tbuist commented:
> We have talked a bit behind the scenes. We think having the progress bars be underneath the balance would be the best place.

> For starters we should start with the progress bars matching what the tooltip shows. So that would not consider spending at all. It would only look at what is the budgeted vs goal. Then categories with long_goal set would look at balance vs goal.

Alright, I'm on board with this. I've updated to make the bars show progress relative to balance. Here's an example that shows behavior with your yearly bill example. I also switched direction of the bars to match @lelemm's suggestion
![image](https://github.com/user-attachments/assets/fb52d064-e0a4-4063-b898-2dbdddba66e3)

Here are the bars now showing on mobile. I added a button in the menu to toggle the preference, same as the desktop version.
![image](https://github.com/user-attachments/assets/63fb4e8b-df43-453d-ac4b-f0efeef76538)
![image](https://github.com/user-attachments/assets/e8a4c08c-4fa4-4f1f-b502-8c2efbbdc5ef)

youngcw commented:
Im a bit confused.  Aren't the progress bars supposed to be based on the template/goals?  Right now they don't consider the templates at all.  Looks like it does consider goals.

tbuist commented:
I haven't been treating the bar as an indicator for the current budgeted amount relative to a #template budget amount. Another example might help me understand how you would like this to behave. I'll use a simple one but you can give a more complex one if it helps. In this scenario for a fresh month, what color should the bar be?
1. Green to represent a positive balance remaining after considering the budgeted and spent amounts
2. Red since my current budgeted amount is less than my #template budget amount
3. Blue since I have a per-month goal (#template) that I haven't budgeted enough for yet

![image](https://github.com/user-attachments/assets/05de84a5-416a-4a95-ae82-36fc1062824e)

youngcw commented:
I thought the idea was to be an indicator of status towards templates/goals.  So use the same information as the goal popover, and same color as the balance coloring, but an analog indicator in addition to the discrete color indication.

tbuist commented:
> So that would not consider spending at all. 

I missed this piece. I don't think it makes sense to introduce an identical analog to what the balance color styling already shows on screen. Perhaps a unique workflow for me, but I prefer to know the status of my spending relative to how much I have to spend for that category, even when that category has a #template.

kopach commented:
If I might add some consideration to this conversation.
Personally, I believe having a longer progress bar as presented by @tbuist [here](https://github.com/actualbudget/actual/pull/4371#issuecomment-2675970503) - **better**. I would just show that bar always if enabled (not on hover) + add some spacing from left/right.

Longer progress bars allow for more information to be shown
1. we can show overbudgeting (with spaces) - IMO super useful
2. we can have **Solid/Striped** colors and much more

please see full description on how YNAB does it https://support.ynab.com/en_us/progress-bars-a-guide-SkDEhot09
![image](https://github.com/user-attachments/assets/72eb673e-e43f-42eb-a4e8-532382d14aaa)

---

Additionally, I believe, we should add toggle option above table to on/off this (specifically to mitigate that cluttering issue mentioned above)
again, example how YNAB does it
![image](https://github.com/user-attachments/assets/7d9057c9-f3b3-41d1-a8e8-e499bff399f0)

Juulz commented:
I have a question. My biggest problem knowing where I am with templates is the "up to" "spend from" situation. Vacations, Medical out-of-pocket sinking fund, and the like.  I have a hard time telling that I have funded the category fully as the balance is usually less than the "up to" amount. Only for some do I want to keep them topped up. Are these progress bars going to help with that?

Derbicus commented:
I really like this feature as depicted here. Having a switch for "savings goals" that inverts the behavior (spread out to a maturation date) from typical draining budgets like "restaurants" in your example would be great. Also a simple button to show or hide allows the option for people to not use them. Perfect.

I want to see this PR get some love and not go stale

tbuist commented:
I’d also like it to get some love too, though I’m not sure who I need to convince (or who gets the final say on behavior) since there have been so many subjective opinions already expressed in this thread. I will wait for a more authoritative comment from the maintainers before continuing to spend my time on it.

youngcw commented:
I think the confusion was that we had thought you were trying to make progress bars for templates and goals exclusively, but you were trying to make progress bars for different things, not just the templates.  The linked feature request was for specifically templates/goals.

tbuist commented:
I see the confusion. My intent was to add status indicators for 2 different types of targets, those being 1) target savings, which are #goal templates (blue bars), and 2) target spending, which is the budget/spent/balance statuses for that month (green/red bars). I agree are two similar but separate concepts and which do not include non-#goal templates, so I understand that it is using the same solution for two different things (though I would suggest the color differentiation solves that).

If #2965 is specifically for #goal templates and not #&NoBreak;2, then my branch contributes too much. The userscript example from last year on the feature request seems to include #&NoBreak;2 as well, though I'm basing that off of the screenshot, so I am not sure if users are assuming that to be a part of the request too. Are you aware of other feature requests that are specific to a category's budget/spent/balance status?

tbuist commented:
Closing PR. I'm sure someone would be able to vibe code a good solution once the behavior is agreed upon.
